src/main/java/httpServer/ListenerAndResponder.java

```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
-            String httpResponse = router.getResponse(request);
-            printServerResponse(httpResponse, clientSocket);
+            HTTPResponse httpResponse = router.getResponse(httpRequest);
+            String responseString = httpResponse.toString();
+            System.err.println(responseString.length());
+            printServerResponse(responseString, clientSocket);
         }
 
     }
 
-    private void printServerResponse(String message, Socket clientSocket) throws IOException {
-            ClientWriteable printer = clientWriterFactory.makePrinter(clientSocket);
-            System.out.println(message);
-            printer.println(message);
-
+    private void printServerResponse(String httpResponse, Socket clientSocket) throws IOException {
+        ClientWriteable printer = clientWriterFactory.makePrinter(clientSocket);
+        System.out.println(httpResponse);
+        printer.print(httpResponse);
+        printer.close();
     }
 
-    private String readAllLines(ClientReadable bufferedReader)  {
-        String line;
-        StringBuilder stringBuilder = new StringBuilder();
+    private HTTPRequest readAllLinesAndReturnHttpRequest(ClientReadable bufferedReader)  {
+
+        String character;
+        RequestBuilder requestBuilder = new RequestBuilder();
+        StringBuilder headers = new StringBuilder();
+        StringBuilder body = new StringBuilder();
+
         try {
-            while ((line = bufferedReader.readLine()) != null)
-            {
-                stringBuilder.append(line);
-                stringBuilder.append(CRLF);
-                if (line.equals("")) { break; }
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
+                    break;
+                }
             }
-            return stringBuilder.toString();
+
         } catch (IOException e) {
             e.printStackTrace();
-            return stringBuilder.toString();
+            return requestBuilder.build();
+        }
+
+        if (checkIfHeadersIndicateARequestBody(headers)) {
+            int bodyLength = getBodyLength(headers);
+            try {
+                while ((character = bufferedReader.read()) != null) {
+                    if (java.util.Objects.equals(bodyLength, 0)) {
+                        break;
+                    }
+                    body.append(character);
+                    if (body.length() == bodyLength) {
+                        break;
+                    }
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+                return requestBuilder.build();
+            }
+        }
+        return turnStringBuilderIntoHttpRequest(headers, body, requestBuilder);
+    }
+
+    private HTTPRequest turnStringBuilderIntoHttpRequest(StringBuilder headers, StringBuilder body, RequestBuilder requestBuilder) {
```

Comment:
I like how you're moving to more descriptive variable names but I think you have a tendency to get bogged down with describing the exact implementation details rather than the abstraction. What's important to relay here? The fact that this method is using a `StringBuilder` or that this method is constructing a `HTTPRequest`?

Reply:
I guess it's that the method is constructing a `HTTPRequest` is the most important thing, and the fact that it takes in a  `StringBuilder` is obvious when you look at the parameters... Aha. So that can make the method name more terse~